### ISSUE TITLE
StackMetadata causes CloudFormation to hang on creation & deletion, v1.18.4

### ISSUE THREAD

**Issue by ceigey** [sst/sst#2222]
Hi all,
We have been trialling Serverless Stack with a lot of excitement but today we hit a stumbling block.

We were on v1.15.16, and then when I tried to use the `bind` method, I realised we were too many versions behind, so we bumped the version all the way to latest (v1.18.4).

Then when we tried to do `npx sst start` to continue development, the deployment hung indefinitely (~ 30min) on a small stack. One item was permanently stuck on `CREATE_IN_PROGRESS`, which was `StackMetadata`.

Eventually I tried to delete the stack from CloudFormation and start again from scratch. But when I tried this, the deletion failed with the following message:

- `The following resource(s) failed to delete: [StackMetadata].`

I then noticed that v1.18.4 had an update with the message "Fix throttled error on updating stack metadata", but I don't know if the two are related. I tried to find the corresponding AWS resource but wasn't sure where to look, I had the impression it was somehow related to Stack Manager but I couldn't find anything obvious with a name like `localdev-myapp-MyStack-....`.

- https://github.com/serverless-stack/sst/commit/8e4ffda9dd7f97df470a614c9e2c5882a50c78ae - the relevant(?) commit

Anyway, I ended up able to delete the stack again eventually, somehow the stack metadata was able to be deleted the second time. I've rolled back the app version to v1.18.2 and I am able to deploy again (taking ~3min).

Thank you for any assistance and information on what might have gone wrong!

---

Confounding factors:

- We were trying out EventBus rules at the time; at one point one of the EventBus rules reported to CloudFormation that it needed manual resource creation, which I thought was unusual, but it was eventually created.
- I deleted one old stack which was removed from the app, and renamed the remaining stack
- I changed two resource names (1x EventBus and 1x Api construct)
- It's Black Friday, maybe AWS was just having a rough day

For what it's worth, the stack consisted of the following:

```typescript
export function MainStack({ stack }: StackContext) {
  const bus = new EventBus(stack, "MainBus");

  const queue = new Queue(stack, "TestQueue", {
    consumer: "functions/debuglogger/lambda.handler",
  });

  const api = new Api(stack, "MainApi", {
    defaults: {
      function: {
        environment: {
          MAIN_BUS_NAME: bus.eventBusName,
        },
      },
    },
  });

  bus.addRules(stack, {
    testRule: {
      pattern: { source: ["some.anonymised.event.source"] },
      targets: {
        queue: queue,
        debuglogger: "functions/debuglogger/lambda.handler",
      },
    },
  });

  api.addRoutes(stack, {
    "POST /adaptwest/reportbridge": {
      function: {
        handler: "functions/oneofourclients/reportbridge/lambda.handler",
      },
    },
  });

  stack.addOutputs({
    ApiEndpoint: api.url,
    EventBusName: bus.eventBusName,
  });

  return { bus, api };
}
```

**Comment by fwang** [comment]
@ceigey lemme give it a try.

**Comment by fwang** [comment]
@ceigey how many stacks r there in the app? And have you created any secrets ie. using `sst secrets`, if so how many?

**Comment by ceigey** [comment]
> @ceigey how many stacks r there in the app? And have you created any secrets ie. using `sst secrets`, if so how many?

Hi @fwang apologies for the late reply! Regarding stacks, we had 3x overall:
- first we had "TestStack"
- then we added "MyStack" to test having 2x stacks
- then we renamed "MyStack" to "[Business Name]Stack"*
- then we deleted "TestStack"

So I think in effect, from a CloudFormation POV, we basically orphaned 2x stacks and created a copy of one with the exact same resources but a new stack name based on the stack function being renamed.

With `sst secrets`, unfortunately we didn't commit this part, but we seem to have added either `Config.Parameter` or the equivalent from Secrets. I think it was just `Config.Parameter` because I can see we _did_ commit the code for using the Node helper `Config` from one of the lambdas.

We definitely didn't use the secrets cli at all though.

*we're rebranding next year so I want to keep the name a secret 😅

**Comment by fwang** [comment]
Thanks for the details @ceigey.

I'm not able to reproduce the error following these steps. I just created a new project with v1.15.16, deployed, updated to v1.18.4, deployed again.

Can you try deploying again, and if metadata gets stuck again:
1. go into your AWS CloudWatch console
2. search for the log group prefix with `/aws/lambda/{STACK_NAME}_ CustomResourceHandler`
3. check the most recent log streams in the log group, you should see an invocation that had the long duration and caused the timeout 
Could u share the logs from it?

**Comment by ceigey** [comment]
Thank you for the instructions – **I haven't redeployed with v1.18.4 yet** – but I can now look at the existing logs.

(I had redeployed with v1.18.2 without issues)

---

So first back to the CloudFormation Event logs,  I found this for `localdev-[business]-[Business]Stack`:

At `2022-11-25 15:43:30 UTC+1030`, StackMetadata = `CREATE_IN_PROGRESS`

At `2022-11-25 16:00:23 UTC+1030`,  localdev-[business]-[Business]Stack = `DELETE_IN_PROGRESS` - _so actually we must have deleted after 15min, and I got the timeline confused!_

At `2022-11-25 16:00:24 UTC+1030`, StackMetadata = `CREATE_FAILED`.

At `2022-11-25 17:00:59 UTC+1030`, StackMetadata = `DELETE_FAILED` with the following message:

> CloudFormation did not receive a response from your Custom Resource. Please check your logs for requestId [7736c9f8-ae0e-4767-b9ca-8ae3e7391ca0]

I've downloaded all the logs for the last 4 weeks with the keyword `CustomResourceHandler` and I can't find this requestID, so I'm not sure which log group CloudFormation wants me to check :-/

---

Now back to CloudWatch logs for `localdev-[business]-[Business]Stack-CustomResourceHandler[....]`

At `2022-11-25T15:43:31.887+10:30`
& `2022-11-25T15:43:32.106+10:30`:

```
2022-11-25T05:13:31.887Z	undefined	ERROR	Uncaught Exception 	{
    "errorType": "Runtime.HandlerNotFound",
    "errorMessage": "index.handler is undefined or not exported",
    "stack": [
        "Runtime.HandlerNotFound: index.handler is undefined or not exported",
        "    at Object.UserFunction.js.module.exports.load (file:///var/runtime/index.mjs:1038:15)",
        "    at async start (file:///var/runtime/index.mjs:1200:23)",
        "    at async file:///var/runtime/index.mjs:1206:1"
    ]
}
```

At `2022-11-25T15:43:32.140+10:30`

```
START RequestId: d150008f-d018-4f1b-bdca-cb11a27f3248 Version: $LATEST
Unknown application error occurred
Runtime.HandlerNotFound
END RequestId: d150008f-d018-4f1b-bdca-cb11a27f3248
```

These errors repeat in a cycle about 4 to 5 times until `2022-11-25T15:46:32.945+10:30`

--- 

Then, under a different log stream for the same log group
From `2022-11-25T16:00:44.470+10:30` 
Until `2022-11-25T16:03:34.564+10:30`

The same cycle reoccurs another 4 to 5 times, same kind of errors etc.

---

Likewise, CloudWatch logs for `localdev-[business]-DemoStack-CustomResourceHandler[....]`

_It looks like we renamed `TestStack` to `DemoStack at some point before deleting it, we also had given it a custom domain at this point and that's perhaps where things started getting weird..._

From `2022-11-25T15:12:45.030+10:30`
Until `2022-11-25T15:15:40.074+10:30`

There's an identical cycle of 4 to 5 uncaught exceptions & failed requests.

But this has no corresponding `StackMetadata` issue in CloudFormation, so I don't know what the connection could be.

**Comment by ceigey** [comment]
My own thought looking at these logs and our git history is maybe we tried too many "risky" things at once and put the CustomResourceHandler into an inconsistent state. E.g., renaming stacks, adding custom domain URLs, deleting stacks, etc. 

And then maybe it just so happened that CloudFormation was waiting for a response from a CustomResourceHandler for `StackMetadata` that either no longer existed or no longer responding.

In which case then I wonder if it's an underlying CloudFormation issue with the Custom Resources?

**Comment by fwang** [comment]
Appreciate the logs!!

> index.handler is undefined or not exported
Let me test something on my end.

**Comment by ceigey** [comment]
Oh no, déjà vu... Started a new app with create-sst on v1.18.4, minimal go starter template.

Unable to finish creating `StackMetadata`, started creating it at 1:20pm, still hasn't finished by 2:12pm, so 52 minutes and still counting 😕

Alternatively, it could also be `apiRouteGETIntegrationGETPermission*` or something like that.

There is an entry for `CustomResourceHandlerE8FB56BA` with `CREATE_COMPLETE`, so that part seems OK.

Screenshot from 2:12pm.

![Screen Shot 2022-12-01 at 2 11 24 pm](https://user-images.githubusercontent.com/23680199/204960635-2c2840b6-e440-4fe8-939f-312b4e2d1b94.png)

**Comment by fwang** [comment]
I might've found the cause. Try emptying the S3 bucket that starts with `cdk-hnb659fds-assets-`, and then deploy again (after current deployment finishes).

**Comment by ceigey** [comment]
Thanks @fwang ! Reattempting the deployment at 3:41pm (UTC+1030), looks like StackMetadata works now. Deployment completed in 3 minutes (3:44pm).

For future reference, are there any negative side-effects to deleting that bucket, especially if I have other CDK apps or stacks running?

---

I can also confirm with my terminal logs that this StackMetadata creation will time-out by the 1hr mark and return an error.

Then, on CloudFormation console, it will automatically try a rollback and again it will time-out after 1hr of StackMetadata being stuck on `DELETE_IN_PROGRESS`.


Terminal logs at 2:20pm:

```
Checking deploy status...
Checking deploy status...
Checking deploy status...
Checking deploy status...
Checking deploy status...
christianm-roberto-MainStack | CREATE_FAILED | Custom::StackMetadata | StackMetadata | CloudFormation did not receive a response from your Custom Resource. Please check your logs for requestId [d9a850cb-9adf-46ab-88f1-052983be4b67]. If you are using the Python cfn-response module, you may need to update your Lambda function code so that CloudFormation can attach the updated version.
christianm-roberto-MainStack | ROLLBACK_IN_PROGRESS | AWS::CloudFormation::Stack | christianm-roberto-MainStack | The following resource(s) failed to create: [StackMetadata]. Rollback requested by user.
christianm-roberto-MainStack | DELETE_IN_PROGRESS | AWS::SSM::Parameter | apiParameterurl1D4D90E2
```

Screenshot taken at 3:38pm, rollback failing after another hour.
![Screen Shot 2022-12-01 at 3 38 32 pm](https://user-images.githubusercontent.com/23680199/204970495-a34d5455-78c3-4054-b9e6-efed085d9b9e.png)

From the terminal logs:

```
christianm-roberto-MainStack | DELETE_COMPLETE | AWS::IAM::Role | CustomResourceHandlerServiceRole41AEC181
christianm-roberto-MainStack | ROLLBACK_FAILED | AWS::CloudFormation::Stack | christianm-roberto-MainStack | The following resource(s) failed to delete: [StackMetadata]. 

 ❌  christianm-roberto-MainStack failed: CloudFormation did not receive a response from your Custom Resource. Please check your logs for requestId [d9a850cb-9adf-46ab-88f1-052983be4b67]. If you are using the Python cfn-response module, you may need to update your Lambda function code so that CloudFormation can attach the updated version.
```

---

After emptying the bucket, a positive sign! `StackMetadata` = `CREATE_COMPLETE`!

![Screen Shot 2022-12-01 at 3 43 41 pm](https://user-images.githubusercontent.com/23680199/204971178-b534fc39-4ea6-41ee-9c03-8fcaca0c3766.png)

**Comment by michaelleeallen** [comment]
+1 Having the same issue, or at least the same result: running `npm run dev` with v1.18.4 _and_ 1.18.3. It seems to hang on the database stack while setting up the migration script(see image). If I do `npm run remove`, everything gets destroyed except the database stack because, again, the migration script.

<img width="828" alt="Screen Shot 2022-12-16 at 9 00 08 AM" src="https://user-images.githubusercontent.com/62591/208117818-8c9b7bd0-da28-48b8-8c3c-be6230c28226.png">

**Comment by CodyDunlap** [comment]
I have had the same issue with the upgrade to 1.18.4 from 1.15.13. Our app has ~30 stacks in it but I haven't been able to identify any correlation to the types of constructs being created and this issue.

I haven't tried emptying the S3 bucket as suggested yet. Like @ceigey, I'm curious if there are side effects to emptying that bucket? The SST app encountering this issue is not the only SST app running in that account so I'd be deleting files associated with other apps as well.

**Comment by fwang** [comment]
@ceigey quick update on the issue. We finally found the root cause. It's a CDK bug where the Lambda zip gets corrupted when uploading multiple concurrently. We've put in a fix in SST v2. Will cut a release candidate soon.

@michaelleeallen I'm not sure if it's caused by the upload issue. Can you go to the AWS Lambda console, find the db migration handler, and export the code? It'd help clarifying if the upload is corrupted.

@CodyDunlap r u having the issue w/ metadata or db migration? Btw, I gave it some test, emptying the CDK assets bucket does not affect deployed apps. So safe to deploy.

**Comment by thdxr** [comment]
This is fixed in v2 but wondering if we need to backport the fix?

**Comment by mpb-prophero** [comment]
If you compare the diff between `1.18.2` and `1.18.4` [here](https://github.com/serverless-stack/sst/compare/v1.18.2...v1.18.4#diff-54dea6b90633ea0377cf7b41eb23693c4b45f9ce581160d20a121a7f3ba02f16)

It seems that the culprit for the hang is the `maxRetries: 1000`:

```
  maxRetries: 1000,
  retryDelayOptions: {
    customBackoff: () => 3000,
  },
```

at least partially anyway.

As mentioned earlier in this thread - we rolled back to version `1.18.2` and the problem has disappeared.

I used the CLI command `sst update 1.18.2` to get the correct dependencies. Pretty sure I nuked  the `node_modules` directory (`rm -rf node_modules`) and the `package-lock.json` file first though.
